libs/contrib/Data/Fun/Extra.idr

```diff
@@ -102,3 +102,14 @@ introduceWitness : {0 r : Rel ts} -> (witness : HVect ts) ->
                    uncurry {ts} r witness -> Ex ts r
 introduceWitness []             f = f
 introduceWitness (w :: witness) f = (w ** introduceWitness witness f)
+
+----------------------------------------------------------------------
+public export
+data FunVect : (ts, ss : Vect n Type) -> Type where
+  Nil  : FunVect [] []
+  (::) : (f : t -> s) -> FunVect ts ss -> FunVect (t::ts) (s::ss)
+
```

Comment:
Isn't this `HVect (zipWith (\ a, b => a -> b) ts ss)`?

Reply:
For example:
```idris
export
Poset Nat (<<=) where
  antisymmetric a b = precompose [LTEbyPlus a b, LTEbyPlus b a]
                               $ antisymmetric a b
```
gives:
```

```
     While processing right hand side of antisymmetric. Can't solve constraint
     between: [(a <<= b -> LTE a b), (b <<= a -> LTE b a)] and zipWith (\a, b => a -> b) ?ts ?ss.
     
     Data/Nat/Order/Strict.idr:108:34--108:64
          |
      108 |   antisymmetric a b = precompose [LTEbyPlus a b, LTEbyPlus b a]
```
indexing by computations is something I really interested in, but this much indexing might be a little too interesting at the moment.

Reply:
In this case what about using:

```idris
data Pointwise : (a -> b -> Type) -> (ts : Vect n a) -> (ss : Vect n b) -> Type where
  Nil  : Pointwise [] []
  (::) : (f : r t s) -> Pointwise r ts ss -> Pointwise r (t::ts) (s::ss)
```

It would be nice not to have to rely on too many ad-hoc data structures
(e.g. how `HVect` is essentially `Data.Vect.All id`...)

Reply:
I see. 
I get different errors this time:
```idris
public export
data Pointwise : (r : a -> b -> Type) -> (ts : Vect n a) -> (ss : Vect n b) -> Type where
  Nil  : Pointwise r [] []
  (::) : (f : r t s) -> Pointwise r ts ss -> Pointwise r (t::ts) (s::ss)

public export
Arrow : Type -> Type -> Type
Arrow a b = a -> b

public export
precompose : Pointwise Arrow ts ss -> Fun ss cod -> Fun ts cod
precompose [] h = h
precompose (f :: fs) h = \x => precompose fs (h ?hole)
```

```
Data/Nat/Order/Strict.idr line 109 col 33:
     While processing right hand side of antisymmetric. Can't find an implementation for Poset Nat ?po.
     
     Data/Nat/Order/Strict.idr:109:34--109:51
          |
      109 |                                $ antisymmetric a b
          |                                  ^^^^^^^^^^^^^^^^^
```
It's entirely possible there's another instability elsewhere in the system.

Reply:
Also, it's clear to me that neither `Fun` nor `All ts cod` are the final say --- we should one day be able to deal with telescopes. So whatever we're currently doing is already 'ad-hoc'

Reply:
Thanks, that works now. I have to include the annotation on `(::)`, otherwise the type-checker fails on the second clause of `precompose`. Unification error?

```idris
public export
data Pointwise : (r : a -> b -> Type) -> (ts : Vect n a) -> (ss : Vect n b) -> Type where
  Nil  : Pointwise r [] []
  (::) : {0 ss, ts : Vect n Type} -> 
         (f : r t s) -> Pointwise r ts ss -> Pointwise r (t::ts) (s::ss)

public export
precompose : Pointwise (\a,b => a -> b) ts ss -> Fun ss cod -> Fun ts cod
precompose [] h = h
precompose (f :: fs) h = \x => precompose fs (h (f x))
```

Reply:
@edwinb  just in case

Reply:
There's no hope of getting good inference for telescopes in their full
generality but stuff like `Fun` is very amenable to it. Cf. [self-promotion](http://gallais.github.io/pdf/tyde19.pdf).

Reply:
I know the self-promoted paper: I wish we had a telescopic version of it.

Putting inference aside, it would be nice to have telescopes in full generality with type ascription. E.g., I'm happy to annotate with the telescope, if it gives me the rest for free. It's how to encode telescopes that I'm interested in.

Reply:
It's not too hard. There is quite a bit of wiggle room depending on whether you
want them left or right nested or whether you want full trees. If you want full
trees then you need IR (cf. [this example in Agda](https://github.com/liamoc/dddp/pull/1)).

Anyway, we should probably move this discussion somewhere else. :D

Reply:
Happy to! Is this final issue resolved?

Reply:
I think so! Eventually we should move `Pointwise` to `Data.Vect.Quantifiers` in `base`
but we don't have any interesting results to attach to it at the moment so it's not
urgent.